src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexCache.cs

```diff
@@ -0,0 +1,192 @@
+
+        /// <summary>The number of records to be batched up and written out at once during cache flushes.</summary>
+        private const int BatchSize = 100;
+
+        /// <summary>Maximum number of items (addresses and balance changes) in the cache.</summary>
+        public readonly int MaxItems;
+
+        /// <summary>Number of address index records & associated balance changes present in the cache.</summary>
+        private int itemCount;
+
+        /// <summary>All access to the cache must be protected with <see cref="lockObj"/>.</summary>
+        private readonly Dictionary<string, AddressIndexerData> cachedAddresses;
+
+        /// <summary>The dirty address cache contains the set of addresses that have been
+        /// modified since their inclusion into the cache itself. This speeds up the
+        /// flushing process as only dirty addresses need to be flushed.</summary>
+        /// <remarks>All access to the dirty address cache must be protected with <see cref="lockObj"/>.</remarks>
+        private readonly HashSet<string> dirtyAddresses;
+
+        private object lockObj;
+
+        private readonly LiteDatabase db;
+
+        private readonly LiteCollection<AddressIndexerData> addressIndexerData;
+
+        public AddressIndexCache(LiteDatabase db, string addressIndexerCollectionName)
+        {
+            this.lockObj = new object();
+            this.db = db;
+            this.addressIndexerData = this.db.GetCollection<AddressIndexerData>(addressIndexerCollectionName);
+            this.addressIndexerData.EnsureIndex("BalanceChangedHeightIndex", "$.BalanceChanges[*].BalanceChangedHeight", false);
+
+            this.MaxItems = AddressIndexCacheMaxItemsDefault;
+            this.itemCount = 0;
+
+            this.dirtyAddresses = new HashSet<string>();
+
+            // The cache is initially empty. A possible improvement could be
+            // persisting it to disk on node shutdown.
+            this.cachedAddresses = new Dictionary<string, AddressIndexerData>();
+        }
+
+        public void AddToCache(AddressIndexerData indexData)
+        {
+            lock (this.lockObj)
+            {
+                // TODO: Have a mapping of last accessed timestamps so items can age out too
+                // TODO: LRU concept - use the dictionary as a lookup for index positions into an array-like data structure
+
+                // Don't bother adding the entry if it exists already.
+                if (this.cachedAddresses.ContainsKey(indexData.Address))
+                    return;
+
+                // Now check if any entries need to be evicted.
+
+                // The size is the base record plus each balance change.
+                // We need to count it this way because an address may have very
+                // few changes and should therefore have less weight than one
+                // with numerous changes.
+                int newItemCount = indexData.BalanceChanges.Count + 1;
```

Comment:
Explain again the +1 here?

Reply:
I regard the index data instance itself as being '1 unit' when calculating the total size of the cache, as it consumes a small quantity of RAM even if there are no balance changes associated with it.